contracts/PionexContract.sol

```diff
@@ -0,0 +1,632 @@
+                    tokenOut: address(_settlement.takerToken),
+                    tokenOutAmountMin: _settlement.protocolOutMinimum,
+                    path: path,
+                    to: address(this),
+                    deadline: _settlement.expiry
+                })
+            );
+        }
+
+        _settlement.makerToken.safeApprove(_settlement.protocolAddress, 0);
+    }
+
+    /// @inheritdoc IPionexContract
+    function cancelLimitOrder(PionexContractLibEIP712.Order calldata _order, bytes calldata _cancelOrderMakerSig) external override onlyUserProxy nonReentrant {
+        require(_order.expiry > uint64(block.timestamp), "LimitOrder: Order is expired");
+        bytes32 orderHash = getEIP712Hash(PionexContractLibEIP712._getOrderStructHash(_order));
+        bool isCancelled = LibOrderStorage.getStorage().orderHashToCancelled[orderHash];
+        require(!isCancelled, "LimitOrder: Order is cancelled already");
+        {
+            PionexContractLibEIP712.Order memory cancelledOrder = _order;
+            cancelledOrder.takerTokenAmount = 0;
+
+            bytes32 cancelledOrderHash = getEIP712Hash(PionexContractLibEIP712._getOrderStructHash(cancelledOrder));
+            require(isValidSignature(_order.maker, cancelledOrderHash, bytes(""), _cancelOrderMakerSig), "LimitOrder: Cancel request is not signed by maker");
+        }
+
+        // Set cancelled state to storage
+        LibOrderStorage.getStorage().orderHashToCancelled[orderHash] = true;
+        emit OrderCancelled(orderHash, _order.maker);
+    }
+
+    /* order utils */
+
+    function _validateOrder(
+        PionexContractLibEIP712.Order memory _order,
+        bytes32 _orderHash,
+        bytes memory _orderMakerSig
+    ) internal view {
+        require(_order.expiry > uint64(block.timestamp), "LimitOrder: Order is expired");
+        bool isCancelled = LibOrderStorage.getStorage().orderHashToCancelled[_orderHash];
+        require(!isCancelled, "LimitOrder: Order is cancelled");
+
+        require(isValidSignature(_order.maker, _orderHash, bytes(""), _orderMakerSig), "LimitOrder: Order is not signed by maker");
+    }
+
+    function _validateOrderTaker(PionexContractLibEIP712.Order memory _order, address _taker) internal pure {
+        if (_order.taker != address(0)) {
+            require(_order.taker == _taker, "LimitOrder: Order cannot be filled by this taker");
+        }
+    }
+
+    function _quoteOrderFromMakerToken(
+        PionexContractLibEIP712.Order memory _order,
+        bytes32 _orderHash,
+        uint256 _makerTokenAmount
+    )
+        internal
+        view
+        returns (
+            uint256,
+            uint256,
+            uint256
+        )
+    {
+        uint256 makerTokenFilledAmount = LibOrderStorage.getStorage().orderHashToMakerTokenFilledAmount[_orderHash];
+
+        require(makerTokenFilledAmount < _order.makerTokenAmount, "LimitOrder: Order is filled");
```

Comment:
I'm pondering if an order could be filled twice, since by trader checks maker token filled, but by protocol checks taker token filled. 

In other words, an order could be first fully filled by trader, and it would only affect `orderHashToMakerTokenFilledAmount` record. And then the order could be filled again by protocol, since `orderHashToTakerTokenFilledAmount` is still `0` now, and it could pass the quote check phase.

Not sure if my understanding is correct 🤔

Reply:
Fixed in f8b2d26